hail/python/hailtop/saige/variant_chunk.py

```diff
@@ -0,0 +1,170 @@
+                chunks.append(VariantChunk(interval))
+                first_rank = cur_rank
+                first_locus = cur_locus
+
+        interval = hl.Interval(
+            first_locus,
+            cur_locus,
+            includes_start=True,
+            includes_end=True,
+        )
+        chunks.append(VariantChunk(interval))
+
+    return chunks
+
+
+def create_chunks_by_group(
+    variants: hl.Table, group: hl.ArrayExpression, max_count_per_chunk: int, max_span_per_chunk: int
+) -> List[VariantChunk]:
+    assert require_row_key_variant(variants, 'saige')
+
+    rg = variants.locus.dtype.reference_genome
+
+    variants = variants.select(group=group).explode('group')
+
+    group_metadata = variants.aggregate(
+        hl.agg.group_by(
+            variants.group.group,
+            hl.struct(
+                contig=hl.array(hl.agg.collect_as_set(variants.locus.contig)),
+                start=hl.agg.min(variants.locus.position),
+                end=hl.agg.max(variants.locus.position),
+                count=hl.agg.count(variants.locus),
+            ),
+        )
+    )
+
+    group_metadata = sorted(list(group_metadata.items()), key=lambda x: (x.contig, x.start))
+
+    def variant_chunk_from_groups(groups: List[hl.Struct]) -> VariantChunk:
+        contig = groups[0].contig
+        start = min(g.start for g in groups)
+        end = max(g.end for g in groups)
+        return VariantChunk(
+            hl.Interval(
+                hl.Locus(contig, start, reference_genome=rg),
+                hl.Locus(contig, end, reference_genome=rg),
+                includes_start=True,
+                includes_end=True,
+            ),
+            [g.group for g in groups],
+        )
+
+    chunks = []
+
+    groups = [group_metadata[0]]
+    current_count = 0
+
+    for group, metadata in group_metadata[1:]:
+        first_group = groups[0]
+        if (
+            (current_count + group.count > max_count_per_chunk)
+            or (group.contig != first_group.contig)
+            or (group.end - first_group.end > max_span_per_chunk)
+        ):
+            chunks.append(variant_chunk_from_groups(groups))
+            current_count = 0
+
+        groups.append(group)
+        current_count += group.count
+
+    chunks.append(variant_chunk_from_groups(groups))
+
+    return chunks
+
+
+def create_variant_chunks(
+    dataset: hl.MatrixTable,
+    *,
+    group: Optional[hl.ArrayExpression] = None,
+    max_count_per_chunk: int = 5000,
+    max_span_per_chunk: int = 5_000_000,
+) -> List[VariantChunk]:
+    if group is not None:
+        assert dataset == group._indices.source
+    require_row_key_variant(dataset, 'saige')
+
+    variants = dataset.rows()
+
+    if group is None:
+        return create_chunks_by_contig(variants, max_count_per_chunk, max_span_per_chunk)
+    return create_chunks_by_group(variants, group, max_count_per_chunk, max_span_per_chunk)
+
+
+class VariantChunks:
```

Comment:
This class is just a `List[VariantChunk]`, I think it's better to just pass around the List itself